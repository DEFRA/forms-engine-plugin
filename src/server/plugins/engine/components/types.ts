export interface Label {
  text: string
  classes?: string
  html?: string
  isPageHeading?: boolean
}

export interface Content {
  title?: string
  text: string
  condition?: string
}

export type ListItemLabel = Omit<Label, 'text' | 'isPageHeading'>

export interface ListItem {
  text?: string
  value: string | boolean | number
  hint?: {
    html: string
  }
  checked?: boolean
  selected?: boolean
  label?: ListItemLabel
  condition?: string
}

// TODO: Break this down for each component (Same as model/Component).
export interface ViewModel {
  label?: Label
  type?: string
  id?: string
  name?: string
  value?: any // TODO
  hint?: {
    html: string
  }
  classes?: string
  condition?: string
  errorMessage?: {
    text: string
  }
  summaryHtml?: string
  html?: any // TODO
  attributes: {
    autocomplete?: string
    maxlength?: number
    multiple?: string
    accept?: string
    step?: string
  }
  content?: Content | Content[] | string
  rows?: number
  items?: ListItem[]
  fieldset?: {
    legend?: Label
  }
  children?: ComponentCollectionViewModel
  autocomplete?: string
}

export type MultilineTextFieldViewModel = {
  maxlength?: number
  isCharacterOrWordCount: boolean
  maxwords?: number
} & ViewModel

export type ComponentCollectionViewModel = {
  type: string
  isFormComponent: boolean
  model: ViewModel
}[]

export type DataType =
  | 'list'
  | 'text'
  | 'date'
  | 'monthYear'
  | 'number'
  | 'file'
