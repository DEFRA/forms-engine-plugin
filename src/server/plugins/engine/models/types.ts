import {
  type ConditionWrapper,
  type DatePartsFieldComponent,
  type FileUploadFieldComponent,
  type FormComponentsDef,
  type FormDefinition,
  type InputFieldsComponentsDef,
  type Item,
  type MonthYearFieldComponent,
  type NumberFieldComponent,
  type Section,
  type SelectionComponentsDef
} from '@defra/forms-model'
import { type Expression } from 'expr-eval'

import { type ComponentBase } from '~/src/server/plugins/engine/components/ComponentBase.js'
import { type DataType } from '~/src/server/plugins/engine/components/types.js'
import { type PageControllerClass } from '~/src/server/plugins/engine/pageControllers/helpers.js'
import {
  type FileState,
  type FormState,
  type FormStateValue,
  type FormSubmissionState,
  type RepeatState
} from '~/src/server/plugins/engine/types.js'

export type ExecutableCondition = ConditionWrapper & {
  expr: Expression
  fn: (state: FormSubmissionState) => boolean
}

/**
 * Used to render a row on a Summary List (check your answers)
 */
export interface DetailItemBase {
  /**
   * Name of the component defined in the JSON {@link FormDefinition}
   */
  name: ComponentBase['name']

  /**
   * Title of the component defined in the JSON {@link FormDefinition}
   * Used as a human readable form of {@link ComponentBase.name} and HTML content for HTML Label tag
   */
  label: ComponentBase['title']

  /**
   * Path to page excluding base path
   */
  path: PageControllerClass['path']

  /**
   * String and/or display value of a field. For example, a Date will be displayed as 25 December 2022
   */
  value: string

  /**
   * Flag to indicate if field is in error and should be changed
   */
  inError?: boolean

  /**
   * Raw value of a field. For example, a Date will be displayed as 2022-12-25
   */
  rawValue: FormState | FormStateValue

  url: string
  type?: FormComponentsDef['type']
  title: string
  dataType?: DataType
  subItems?: DetailItem[][]
}

export interface DetailItemDate extends DetailItemBase {
  type: DatePartsFieldComponent['type']
  dataType: DataType.Date
  rawValue: FormState | null
}

export interface DetailItemMonthYear extends DetailItemBase {
  type: MonthYearFieldComponent['type']
  dataType: DataType.MonthYear
  rawValue: FormState | null
}

export interface DetailItemSelection extends DetailItemBase {
  type: SelectionComponentsDef['type']
  dataType: DataType.List
  items: DetailItem[]
  rawValue: Item['value'] | Item['value'][] | null
}

export interface DetailItemNumber extends DetailItemBase {
  type: NumberFieldComponent['type']
  dataType: DataType.Number
  rawValue: number | null
}

export interface DetailItemText extends DetailItemBase {
  type: Exclude<
    InputFieldsComponentsDef,
    NumberFieldComponent | FileUploadFieldComponent
  >['type']
  dataType: DataType.Text
  rawValue: string | null
}

export interface DetailItemFileUpload extends DetailItemBase {
  type: FileUploadFieldComponent['type']
  dataType: DataType.File
  rawValue: FileState[] | null
}

export interface DetailItemRepeat extends DetailItemBase {
  rawValue: RepeatState[] | null
}

export type DetailItem =
  | DetailItemDate
  | DetailItemMonthYear
  | DetailItemSelection
  | DetailItemNumber
  | DetailItemText
  | DetailItemFileUpload
  | DetailItemRepeat

/**
 * Used to render a row on a Summary List (check your answers)
 */
export interface Detail {
  name?: Section['name']
  title?: Section['title']
  items: DetailItem[]
}
