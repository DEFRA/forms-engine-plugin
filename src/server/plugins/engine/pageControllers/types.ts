import type * as PageControllers from '~/src/server/plugins/engine/pageControllers/index.js'

export type PageControllerClass = InstanceType<PageControllerType>
export type PageControllerType =
  (typeof PageControllers)[keyof typeof PageControllers]
