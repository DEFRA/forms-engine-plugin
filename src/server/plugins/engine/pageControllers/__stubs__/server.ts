import { type Server } from '@hapi/hapi'

import { type PluginOptions } from '~/src/server/plugins/engine/types.js'
import { type CacheService } from '~/src/server/services/index.js'

export const server: Server = {
  plugins: {
    'forms-engine-plugin': {
      baseLayoutPath: '',
      cacheService: {} as CacheService
    }
  }
} as Server // only mocking out properties we care about;

export const serverWithSaveAndReturn: Server = {
  plugins: {
    ...server.plugins,
    'forms-engine-plugin': {
      ...server.plugins['forms-engine-plugin'],
      saveAndReturn: {
        keyGenerator: jest.fn().mockReturnValue('foobar'),
        sessionHydrator: jest.fn().mockReturnValue({}),
        sessionPersister: jest.fn().mockImplementation(() => Promise.resolve())
      } as Pick<PluginOptions, 'saveAndReturn'>
    }
  }
} as Server // only mocking out properties we care about
