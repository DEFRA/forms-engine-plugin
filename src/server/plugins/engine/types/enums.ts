export enum UploadStatus {
  initiated = 'initiated',
  pending = 'pending',
  ready = 'ready'
}

export enum FileStatus {
  complete = 'complete',
  rejected = 'rejected',
  pending = 'pending'
}

export enum FormAdapterSubmissionSchemaVersion {
  V1 = 1
}
