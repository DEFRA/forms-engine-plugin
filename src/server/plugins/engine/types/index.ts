export type {
  CheckAnswers,
  ErrorMessageTemplate,
  ErrorMessageTemplateList,
  FeaturedFormPageViewModel,
  FileState,
  FilterFunction,
  FormAdapterSubmissionMessage,
  FormAdapterSubmissionMessageData,
  FormAdapterSubmissionMessageMeta,
  FormAdapterSubmissionMessageMetaSerialised,
  FormAdapterSubmissionMessagePayload,
  FormAdapterSubmissionService,
  FormContext,
  FormContextRequest,
  FormPageViewModel,
  FormPayload,
  FormPayloadParams,
  FormState,
  FormStateValue,
  FormSubmissionError,
  FormSubmissionState,
  FormValidationResult,
  FormValue,
  GlobalFunction,
  ItemDeletePageViewModel,
  OnRequestCallback,
  PageViewModel,
  PageViewModelBase,
  PluginOptions,
  PreparePageEventRequestOptions,
  RepeatItemState,
  RepeatListState,
  RepeaterSummaryPageViewModel,
  SummaryList,
  SummaryListAction,
  SummaryListRow,
  TempFileState,
  UploadInitiateResponse,
  UploadStatusFileResponse,
  UploadStatusResponse
} from '~/src/server/plugins/engine/types.js'

export { FileStatus, UploadStatus } from '~/src/server/plugins/engine/types.js'

export type {
  Detail,
  DetailItem,
  DetailItemBase,
  DetailItemField,
  DetailItemRepeat,
  ExecutableCondition
} from '~/src/server/plugins/engine/models/types.js'

export type {
  BackLink,
  ComponentText,
  ComponentViewModel,
  Content,
  DateInputItem,
  DatePartsState,
  Label,
  ListItem,
  ListItemLabel,
  MonthYearState,
  ViewModel
} from '~/src/server/plugins/engine/components/types.js'

export type { UkAddressState } from '~/src/server/plugins/engine/components/UkAddressField.js'

export type {
  FormParams,
  FormQuery,
  FormRequest,
  FormRequestPayload,
  FormRequestPayloadRefs,
  FormRequestRefs,
  FormResponseToolkit
} from '~/src/server/routes/types.js'

export { FormAction, FormStatus } from '~/src/server/routes/types.js'

export type {
  FormSubmissionService,
  FormsService,
  OutputService,
  RouteConfig,
  Services
} from '~/src/server/types.js'

export type { RichFormValue } from '~/src/server/plugins/engine/outputFormatters/machine/v2.js'

export * from '~/src/server/plugins/engine/types/schema.js'
export { FormAdapterSubmissionSchemaVersion } from '~/src/server/plugins/engine/types/enums.js'
