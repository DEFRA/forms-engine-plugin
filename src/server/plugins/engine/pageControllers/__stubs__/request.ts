import { server } from '~/src/server/plugins/engine/pageControllers/__stubs__/server.js'
import { type FormContextRequest } from '~/src/server/plugins/engine/types.js'
import { type FormRequest } from '~/src/server/routes/types.js'

export function buildFormRequest(
  request: Omit<FormRequest, 'server'>
): FormRequest {
  return {
    ...request,
    server
  } as FormRequest
}

export function buildFormContextRequest(
  request: Omit<FormContextRequest, 'server'>
): FormContextRequest {
  return {
    ...request,
    server
  } as FormContextRequest
}
