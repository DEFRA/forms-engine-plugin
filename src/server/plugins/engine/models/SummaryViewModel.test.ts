import { FORM_PREFIX } from '~/src/server/constants.js'
import {
  FormModel,
  SummaryViewModel
} from '~/src/server/plugins/engine/models/index.js'
import { type DetailItem } from '~/src/server/plugins/engine/models/types.js'
import { SummaryPageController } from '~/src/server/plugins/engine/pageControllers/SummaryPageController.js'
import { buildFormContextRequest } from '~/src/server/plugins/engine/pageControllers/__stubs__/request.js'
import { serverWithSaveAndExit } from '~/src/server/plugins/engine/pageControllers/__stubs__/server.js'
import {
  createPage,
  type PageControllerClass
} from '~/src/server/plugins/engine/pageControllers/helpers/pages.js'
import {
  type FormContext,
  type FormContextRequest,
  type FormState
} from '~/src/server/plugins/engine/types.js'
import definition from '~/test/form/definitions/repeat-mixed.js'
const basePath = `${FORM_PREFIX}/test`

describe('SummaryViewModel', () => {
  const itemId1 = 'abc-123'
  const itemId2 = 'xyz-987'

  let model: FormModel
  let page: PageControllerClass
  let pageUrl: URL
  let request: FormContextRequest
  let context: FormContext
  let summaryViewModel: SummaryViewModel

  beforeEach(() => {
    model = new FormModel(definition, {
      basePath: `${FORM_PREFIX}/test`
    })

    page = createPage(model, definition.pages[2])
    pageUrl = new URL('http://example.com/repeat/pizza-order/summary')

    request = buildFormContextRequest({
      method: 'get',
      url: pageUrl,
      path: pageUrl.pathname,
      params: {
        path: 'pizza-order',
        slug: 'repeat'
      },
      query: {},
      app: { model }
    })
  })

  describe.each([
    {
      description: '0 items',
      state: {
        $$__referenceNumber: 'foobar',
        orderType: 'collection',
        pizza: []
      } satisfies FormState,
      keys: [
        'How would you like to receive your pizza?',
        'Pizzas',
        'How you would like to receive your pizza',
        'Pizzas',
        'Pizza'
      ],
      values: ['Collection', 'Not supplied'],
      answers: ['Collection', ''],
      names: ['orderType', 'pizza']
    },
    {
      description: '1 item',
      state: {
        $$__referenceNumber: 'foobar',
        orderType: 'delivery',
        pizza: [
          {
            toppings: 'Ham',
            quantity: 2,
            itemId: itemId1
          }
        ]
      } satisfies FormState,
      keys: [
        'How would you like to receive your pizza?',
        'Pizza added',
        'How you would like to receive your pizza',
        'Pizzas',
        'Pizza'
      ],
      values: ['Delivery', 'You added 1 Pizza'],
      answers: ['Delivery', 'You added 1 Pizza'],
      names: ['orderType', 'pizza']
    },
    {
      description: '2 items',
      state: {
        $$__referenceNumber: 'foobar',
        orderType: 'delivery',
        pizza: [
          {
            toppings: 'Ham',
            quantity: 2,
            itemId: itemId1
          },
          {
            toppings: 'Pepperoni',
            quantity: 1,
            itemId: itemId2
          }
        ]
      } satisfies FormState,
      keys: [
        'How would you like to receive your pizza?',
        'Pizzas added',
        'How you would like to receive your pizza',
        'Pizzas',
        'Pizza'
      ],
      values: ['Delivery', 'You added 2 Pizzas'],
      answers: ['Delivery', 'You added 2 Pizzas'],
      names: ['orderType', 'pizza']
    }
  ])(
    'Check answers ($description)',
    ({ state, keys, values, names, answers }) => {
      beforeEach(() => {
        context = model.getFormContext(request, state)
        summaryViewModel = new SummaryViewModel(request, page, context)
      })

      it('should add title for each section', () => {
        const [checkAnswers1, checkAnswers2] = summaryViewModel.checkAnswers

        // 1st summary list has no title
        expect(checkAnswers1).toHaveProperty('title', undefined)

        // 2nd summary list has section title
        expect(checkAnswers2).toHaveProperty('title', {
          text: 'Food'
        })
      })

      it('should add summary list for each section', () => {
        expect(summaryViewModel.checkAnswers).toHaveLength(2)

        const [checkAnswers1, checkAnswers2] = summaryViewModel.checkAnswers

        const { summaryList: summaryList1 } = checkAnswers1
        const { summaryList: summaryList2 } = checkAnswers2

        expect(summaryList1).toHaveProperty('rows', [
          {
            key: {
              text: keys[2]
            },
            value: {
              classes: 'app-prose-scope',
              html: values[0]
            },
            actions: {
              items: [
                {
                  classes: 'govuk-link--no-visited-state',
                  href: `${basePath}/delivery-or-collection?returnUrl=${encodeURIComponent(`${basePath}/summary`)}`,
                  text: 'Change',
                  visuallyHiddenText: keys[0]
                }
              ]
            }
          }
        ])

        expect(summaryList2).toHaveProperty('rows', [
          {
            key: {
              text: keys[1]
            },
            value: {
              classes: 'app-prose-scope',
              html: values[1]
            },
            actions: {
              items: [
                {
                  classes: 'govuk-link--no-visited-state',
                  href: `${basePath}/pizza-order/summary?returnUrl=${encodeURIComponent(`${basePath}/summary`)}`,
                  text: 'Change',
                  visuallyHiddenText: 'Pizza'
                }
              ]
            }
          }
        ])
      })

      it('should add summary list for each section (preview URL direct access)', () => {
        request.query.force = '' // Preview URL '?force'
        context = model.getFormContext(request, state)
        summaryViewModel = new SummaryViewModel(request, page, context)

        expect(summaryViewModel.checkAnswers).toHaveLength(2)

        const [checkAnswers1, checkAnswers2] = summaryViewModel.checkAnswers

        const { summaryList: summaryList1 } = checkAnswers1
        const { summaryList: summaryList2 } = checkAnswers2

        expect(summaryList1).toHaveProperty('rows', [
          {
            key: {
              text: keys[2]
            },
            value: {
              classes: 'app-prose-scope',
              html: values[0]
            },
            actions: {
              items: []
            }
          }
        ])

        expect(summaryList2).toHaveProperty('rows', [
          {
            key: {
              text: keys[1]
            },
            value: {
              classes: 'app-prose-scope',
              html: values[1]
            },
            actions: {
              items: []
            }
          }
        ])
      })

      it('should use correct summary labels', () => {
        request.query.force = '' // Preview URL '?force'
        context = model.getFormContext(request, state)
        summaryViewModel = new SummaryViewModel(request, page, context)

        expect(summaryViewModel.details).toHaveLength(2)

        const [details1, details2] = summaryViewModel.details

        expect(details1.items[0]).toMatchObject({
          name: names[0],
          value: answers[0],
          title: keys[2],
          label: keys[0]
        })

        expect(details2.items[0]).toMatchObject({
          name: names[1],
          value: answers[1],
          title: keys[1],
          label: keys[4]
        })

        const snapshot = [
          {
            name: names[0],
            value: answers[0],
            title: keys[2],
            label: keys[0]
          },
          {
            name: names[1],
            value: answers[1],
            title: keys[1],
            label: keys[4]
          }
        ]

        expect(snapshot).toMatchSnapshot()
      })
    }
  )
})

describe('SummaryPageController', () => {
  let model: FormModel
  let controller: SummaryPageController
  let request: FormContextRequest

  beforeEach(() => {
    model = new FormModel(definition, {
      basePath: `${FORM_PREFIX}/test`
    })

    controller = new SummaryPageController(model, definition.pages[2])

    request = {
      method: 'get',
      url: new URL('http://example.com/repeat/pizza-order/summary'),
      path: '/repeat/pizza-order/summary',
      params: {
        path: 'pizza-order',
        slug: 'repeat'
      },
      query: {},
      app: { model },
      server: serverWithSaveAndExit
    }
  })

  describe('Save and Exit functionality', () => {
    it('should show save and exit button on summary page', () => {
      expect(controller.shouldShowSaveAndExit(request.server)).toBe(true)
    })

    it('should handle save and exit from summary page', () => {
      const state: FormState = {
        $$__referenceNumber: 'foobar',
        orderType: 'collection',
        pizza: []
      }

      const context = model.getFormContext(request, state)
      const viewModel = controller.getViewModel(request, context)

      expect(viewModel).toHaveProperty('allowSaveAndExit', true)
    })

    it('should display correct page title', () => {
      const state: FormState = {
        $$__referenceNumber: 'foobar',
        orderType: 'collection',
        pizza: []
      }

      const context = model.getFormContext(request, state)
      const viewModel = controller.getViewModel(request, context)

      expect(viewModel.pageTitle).toBe('Check your answers')
    })
  })
})

describe('SummaryViewModel (summaryPath handling)', () => {
  const itemId = 'pizza-001'
  let model: FormModel
  let basePage: PageControllerClass
  let request: FormContextRequest
  let context: FormContext

  beforeEach(() => {
    model = new FormModel(definition, { basePath })
    basePage = createPage(model, definition.pages[2])
    request = buildFormContextRequest({
      method: 'get',
      url: new URL(`${basePath}/summary`, 'http://example.com'),
      path: `${basePath}/summary`,
      params: { path: 'pizza-order', slug: 'repeat' },
      query: {},
      app: { model }
    })
  })

  it('should use page.getSummaryPath(request) when present', () => {
    const summaryPathFromRequest = `${basePath}/custom-summary-path?param=value`
    const getSummaryPathMock = jest
      .spyOn(basePage, 'getSummaryPath')
      .mockImplementation((req?: FormContextRequest) => {
        return req ? summaryPathFromRequest : `${basePath}/custom-summary-path`
      })

    context = model.getFormContext(request, {
      $$__referenceNumber: 'test1',
      orderType: 'delivery',
      pizza: [{ toppings: 'Cheese', quantity: 1, itemId }]
    })

    const viewModel = new SummaryViewModel(request, basePage, context)
    expect(getSummaryPathMock).toHaveBeenCalledWith(request)

    // SummaryPath should be used as returnPath for ItemField/ItemRepeat
    const pizzaSection = viewModel.details.find((x) => x.title === 'Food') ?? {
      items: []
    }
    const pizzaItem = pizzaSection.items[0]
    // Should get correct returnPath in hrefs
    expect(pizzaItem.href).toMatch(encodeURIComponent(summaryPathFromRequest))

    getSummaryPathMock.mockRestore()
  })

  it('should use summaryPath as returnUrl for ItemField hrefs (non-repeat pages)', () => {
    const summaryPathFromRequest = `${basePath}/custom-summary-path?param=value`
    const getSummaryPathMock = jest
      .spyOn(basePage, 'getSummaryPath')
      .mockImplementation((req?: FormContextRequest) => {
        return req ? summaryPathFromRequest : `${basePath}/custom-summary-path`
      })

    context = model.getFormContext(request, {
      $$__referenceNumber: 'test2',
      orderType: 'delivery',
      pizza: [{ toppings: 'Cheese', quantity: 1, itemId }]
    })

    const viewModel = new SummaryViewModel(request, basePage, context)

    // First summary section contains a non-repeat ItemField (delivery-or-collection)
    const firstSection = viewModel.details[0]
    const firstItem = firstSection.items[0]

    expect(firstItem.href).toContain(encodeURIComponent(summaryPathFromRequest))

    getSummaryPathMock.mockRestore()
  })

  it('should use summaryPath as returnUrl for ItemRepeat subItems hrefs', () => {
    const summaryPathFromRequest = `${basePath}/custom-summary-path?param=value`
    const getSummaryPathMock = jest
      .spyOn(basePage, 'getSummaryPath')
      .mockImplementation((req?: FormContextRequest) => {
        return req ? summaryPathFromRequest : `${basePath}/custom-summary-path`
      })

    context = model.getFormContext(request, {
      $$__referenceNumber: 'test3',
      orderType: 'delivery',
      pizza: [
        { toppings: 'Cheese', quantity: 1, itemId },
        { toppings: 'Ham', quantity: 2, itemId: 'pizza-002' }
      ]
    })

    const viewModel = new SummaryViewModel(request, basePage, context)

    // Find the repeat item (Food section)
    const pizzaSection = viewModel.details.find((x) => x.title === 'Food')
    const repeatItem: DetailItem | undefined = pizzaSection?.items[0]

    // subItems is an array of arrays of ItemField; flatten and check all hrefs
    const subItems = (
      repeatItem && 'subItems' in repeatItem ? repeatItem.subItems : []
    ).flat()
    expect(subItems.length).toBeGreaterThan(0)
    for (const subItem of subItems) {
      expect(subItem.href).toContain(encodeURIComponent(summaryPathFromRequest))
    }

    getSummaryPathMock.mockRestore()
  })
})

describe('SummaryViewModel (summaryPath fallback)', () => {
  const itemId = 'pizza-001'
  let model: FormModel
  let basePage: PageControllerClass
  let request: FormContextRequest
  let context: FormContext

  beforeEach(() => {
    model = new FormModel(definition, { basePath })
    basePage = createPage(model, definition.pages[2])
    request = buildFormContextRequest({
      method: 'get',
      url: new URL(`${basePath}/summary`, 'http://example.com'),
      path: `${basePath}/summary`,
      params: { path: 'pizza-order', slug: 'repeat' },
      query: {},
      app: { model }
    })
  })

  it('should fallback to page.getSummaryPath() when returnPath is undefined for ItemField', () => {
    // Force summaryPath from request to be undefined so items must fall back to their own page.getSummaryPath()
    const getSummaryPathMock = jest
      .spyOn(basePage, 'getSummaryPath')
      .mockImplementation((req?: FormContextRequest) => {
        return req ? (undefined as unknown as string) : `${basePath}/summary`
      })

    context = model.getFormContext(request, {
      $$__referenceNumber: 'test-fallback-1',
      orderType: 'delivery',
      pizza: [{ toppings: 'Cheese', quantity: 1, itemId }]
    })

    const viewModel = new SummaryViewModel(request, basePage, context)

    // First summary section contains a non-repeat ItemField
    const firstSection = viewModel.details[0]
    const firstItem = firstSection.items[0]

    // Should include default summary path in returnUrl
    expect(firstItem.href).toContain(encodeURIComponent(`${basePath}/summary`))

    getSummaryPathMock.mockRestore()
  })

  it('should fallback to page.getSummaryPath() when returnPath is undefined for ItemRepeat and subItems', () => {
    // Force summaryPath from request to be undefined so items must fall back to their own page.getSummaryPath()
    const getSummaryPathMock = jest
      .spyOn(basePage, 'getSummaryPath')
      .mockImplementation((req?: FormContextRequest) => {
        return req ? (undefined as unknown as string) : `${basePath}/summary`
      })

    context = model.getFormContext(request, {
      $$__referenceNumber: 'test-fallback-2',
      orderType: 'delivery',
      pizza: [
        { toppings: 'Cheese', quantity: 1, itemId },
        { toppings: 'Ham', quantity: 2, itemId: 'pizza-002' }
      ]
    })

    const viewModel = new SummaryViewModel(request, basePage, context)

    // Repeat section (Food)
    const pizzaSection = viewModel.details.find((x) => x.title === 'Food')
    const repeatItem: DetailItem | undefined = pizzaSection?.items[0]

    expect(repeatItem?.href ?? '').toContain(
      encodeURIComponent(`${basePath}/summary`)
    )

    // subItems is an array of arrays of ItemField; flatten and check all hrefs
    const subItems = (
      repeatItem && 'subItems' in repeatItem ? repeatItem.subItems : []
    ).flat()

    expect(subItems.length).toBeGreaterThan(0)
    for (const subItem of subItems) {
      expect(subItem.href).toContain(encodeURIComponent(`${basePath}/summary`))
    }

    getSummaryPathMock.mockRestore()
  })
})
