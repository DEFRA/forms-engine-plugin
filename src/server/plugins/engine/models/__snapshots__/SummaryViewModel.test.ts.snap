// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SummaryViewModel Check answers (0 items) should use correct summary labels 1`] = `
[
  {
    "label": "How would you like to receive your pizza?",
    "name": "orderType",
    "title": "How you would like to receive your pizza",
    "value": "Collection",
  },
  {
    "label": "Pizza",
    "name": "pizza",
    "title": "Pizzas",
    "value": "",
  },
]
`;

exports[`SummaryViewModel Check answers (1 item) should use correct summary labels 1`] = `
[
  {
    "label": "How would you like to receive your pizza?",
    "name": "orderType",
    "title": "How you would like to receive your pizza",
    "value": "Delivery",
  },
  {
    "label": "Pizza",
    "name": "pizza",
    "title": "Pizza added",
    "value": "You added 1 Pizza",
  },
]
`;

exports[`SummaryViewModel Check answers (2 items) should use correct summary labels 1`] = `
[
  {
    "label": "How would you like to receive your pizza?",
    "name": "orderType",
    "title": "How you would like to receive your pizza",
    "value": "Delivery",
  },
  {
    "label": "Pizza",
    "name": "pizza",
    "title": "Pizzas added",
    "value": "You added 2 Pizzas",
  },
]
`;
